### DISCUSSION TITLE
Duplicate entries in COO-structure in JSO

### DISCUSSION THREAD

**Discussion by tmigot** [JuliaSmoothOptimizers/Organization#51] [Ideas]
In JSO, we use very often the sparse COO matrix representation (rows, cols, vals). This discussion is to clarify the case of duplicate entries in this representation, i.e. the same pair (row, col) may appear more than once in the representation. The typical package doing this is PDENLPModels.jl.

We realized in a recent issue with @MaxenceGollier
https://github.com/JuliaSmoothOptimizers/PDENLPModels.jl/issues/105 that the presence of duplicate entries is not documented and this may be confusing. For instance, QRMumps.jl assumes that there is no such case.

There is a related discussion in this PR https://github.com/JuliaSmoothOptimizers/SparseMatricesCOO.jl/pull/45 .

So, I believe we should clarify what is the expected behavior in general and in SparseMatricesCOO.
Then, we should also document this behavior at least in the docstring of the NLPModel API.

What do you think?

Cc' @dpo @abelsiqueira @amontoison @geoffroyleconte @paraynaud @d-monnet

**Comment by amontoison** [comment]
Yes, we should add a sentence to clarify that we can have duplicates and also that our sparse COO format is not sorted by rows or columns.
Some routines in CUDA.jl expect the COO to be sorted for example.

**Comment by dpo** [comment]
I agree. It’s important to allow duplicate entries. We’ll need a preprocessing step in QRMumps.jl; I did not realize it doesn’t accept duplicate entries. It’s a bit odd.

**Comment by tmigot** [comment]
Maybe these presolve could be implemented in SparseMatricesCOO.jl ? Can you guys create issues for what you have in mind? Remove duplicate entries, sort, etc?

For the documentation, I can think of two places:
- [NLPModels.jl](https://github.com/JuliaSmoothOptimizers/NLPModels.jl): `jac_structure`, `hess_structure`, `hess_residual_structure` docstrings
- [SparseMatricesCOO](https://github.com/JuliaSmoothOptimizers/SparseMatricesCOO.jl): `SparseMatrixCOO` docstring

**Reply by dpo** [reply]
We don't want to remove duplicates in general. Only in packages where they're not supported. But I wonder if it's intentional that it's not supported in QRMumps.
